core/src/main/clojure/xtdb/operator/scan.clj

```diff
@@ -528,39 +605,363 @@
+                                                      "xt$system_to" (vr/vec->reader
+                                                                      (doto (NullVector. "xt$system_to")
+                                                                        (.setValueCount leaf-row-count)))
+                                                      "xt$valid_from" nil
+                                                      "xt$valid_to" nil
+                                                      (.structKeyReader doc-rdr normalized-name))]
+                           :when rdr]
+                       (.rowCopier rdr
+                                   (case normalized-name
+                                     "_iid" (.writerForName out-rel col-name :i64)
+                                     "xt$system_to" (.writerForName out-rel col-name
+                                                                    [:union #{:null types/temporal-col-type}])
+                                     (.writerForName out-rel col-name)))))
+
+        valid-from-wtrs (vec
+                         (for [col-name col-names
+                               :when (= "xt$valid_from" (util/str->normal-form-str col-name))]
+                           (.writerForName out-rel col-name types/temporal-col-type)))
+
+        valid-to-wtrs (vec
+                       (for [col-name col-names
+                             :when (= "xt$valid_to" (util/str->normal-form-str col-name))]
+                         (.writerForName out-rel col-name types/temporal-col-type)))
+
+        sys-from-wtrs (vec
+                       (for [col-name col-names
+                             :when (= "xt$system_from" (util/str->normal-form-str col-name))]
+                         (.writerForName out-rel col-name types/temporal-col-type)))
+
+        skip-iid-ptr (ArrowBufPointer.)
+        prev-iid-ptr (ArrowBufPointer.)
+        current-iid-ptr (ArrowBufPointer.)
+
+        ;; TODO the ranges could be kept in sorted order which could speed up some checks
+        !ranges (LinkedList.)
+        !new-ranges (LinkedList.)]
+
+    (letfn [(duplicate-ptr [^ArrowBufPointer dst, ^ArrowBufPointer src]
+              (.set dst (.getBuf src) (.getOffset src) (.getLength src)))]
+
+      (reify IntConsumer
+        (accept [_ idx]
+          (when-not (= skip-iid-ptr (.getPointer iid-rdr idx current-iid-ptr))
+            (when-not (= prev-iid-ptr current-iid-ptr)
+              (.clear !ranges)
+              (duplicate-ptr prev-iid-ptr current-iid-ptr))
+
+            (if (= :evict (.getLeg op-rdr idx))
+              (duplicate-ptr skip-iid-ptr current-iid-ptr)
+              (let [sys-from (.getLong sys-from-rdr idx)]
+                (when (<= sys-from system-time)
+                  (case (.getLeg op-rdr idx)
+                    :put
+                    (let [i1 (->Interval (.getLong put-valid-from-rdr idx) (.getLong put-valid-to-rdr idx) sys-from)
```

Comment:
by calling `->Interval` here I _think_ you're boxing up the two values - whereas you don't with `Interval.` - happy to be corrected on that though 🙂